### ISSUE TITLE
Use GHC's AST rather than reparsing the original source code

### ISSUE THREAD

**Issue by ocharles** [ocharles/hlint-source-plugin#3]

**Comment by ndmitchell** [comment]
I suggest not doing anything on this for at least a few weeks. A plan is afoot, but it requires some stuff that isn't yet public.

**Comment by ocharles** [comment]
Well well well! I'll sit tight, feel free to drop me a private email if
there's anything to coordinate on

**Comment by ocharles** [comment]
@ndmitchell any update? I'm keen to push this forward now 8.6 is out

**Comment by ndmitchell** [comment]
Afraid nothing public yet. Note that HLint doesn't compile with 8.6 yet because of some of its dependencies, so want to get that sorted first.

**Comment by ndmitchell** [comment]
To keep you in the loop, I'm hoping to be able to make the plan public in a few weeks. The HLint on 8.6 issue is now fixed, which was the top priority.

**Comment by ndmitchell** [comment]
Plan is that HLint will probably switch to [`ghc-lib`](https://github.com/digital-asset/ghc-lib) as it's parser. This will mean that for a specific release of GHC you'll be able to use it direct via the AST, for all others you'll probably have to reparse.

**Comment by tfausak** [comment]
Didn't HLint 3 switch to using GHC for parsing? Does that mean this is possible now?

**Comment by shayne-fletcher** [comment]
Yes. HLint is using the ghc-8.10 parse tree and the API admits passing one you parsed earlier for linting analysis.

**Comment by tfausak** [comment]
Cool 😄 I made a proof of concept for running HLint as a GHC source plugin without re-parsing modules. Obviously it's very similar to this repository. Check it out! https://github.com/tfausak/splint

**Comment by shayne-fletcher** [comment]
"What a time to be alive" 🤣
